src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js

```diff
@@ -1,154 +1,372 @@
+            brushElem.call(brush.move, null);
+        }
 
-            var y = d3.scaleLinear()
-                .domain([0, 9])
-                .range([height, 0]);
-            svg.append("g")
-                .call(d3.axisLeft(y));
+        const zoomInKeyEvent = function (event) {
+            if ((event.ctrlKey || event.metaKey) && event.key === 'z') {
+                zoomIn();
+                selectedZoomBtn.style.display = "none";
+            }
+        };
 
-            var clip = svg.append("defs").append("svg:clipPath")
-                .attr("id", "clip")
-                .append("svg:rect")
-                .attr("width", width)
-                .attr("height", height)
-                .attr("x", 0)
-                .attr("y", 0);
+        function updateChart() {
+            let s = d3.event.selection;
+            selectedZoomBtn.style.display = "inline-block";
+            selectedZoomBtn.addEventListener("click",zoomIn,true)
+            document.addEventListener('keydown', zoomInKeyEvent,true);
+            extent = s;
+        }
 
-            var color = d3.scaleOrdinal()
-                .domain(colorDomain)
-                .range(colorRange)
+        const endBrushing = function (_) {
+            brushElem.call(brush.move, null);
+            selectedZoomBtn.style.display = "none";
+            selectedZoomBtn.removeEventListener("click",zoomIn,true)
+            document.removeEventListener('keydown', zoomInKeyEvent,true);
+        };
 
-            var brush = d3.brushX()
-                .extent([[0, 0], [width, height]])
-                .on("end", updateChart)
+        document.addEventListener('click'      , endBrushing,false);
+        document.addEventListener('auxclick'   , endBrushing,false);
+        document.addEventListener('contextmenu', endBrushing,false);
+        document.addEventListener('scroll'     , endBrushing,false);
+    }
 
-            var scatter = svg.append('g')
-                .attr("clip-path", "url(#clip)")
+    zoomingHelper(scaleAndAxis, box_width, scatter, points) {
+        scaleAndAxis.xAxis.transition().duration(1000)
+            .call(d3.axisBottom(scaleAndAxis.xScale).ticks(box_width / num_width));
+        scaleAndAxis.yAxis.transition().duration(1000)
+            .call(d3.axisLeft(scaleAndAxis.yScale));
 
-            scatter
-                .selectAll("circle")
-                .data(_data)
+        scatter.selectAll("path")
+            .transition().duration(1000)
+            .attr('transform', d => "translate(" + scaleAndAxis.xScale(d.x) + "," + scaleAndAxis.yScale(d.y) + ")")
+
+        if (points.labels === "visible") {
+            scatter.selectAll("text")
+                .transition().duration(1000)
+                .attr("x", d => scaleAndAxis.xScale(d.x) + lbl_padding_x)
+                .attr("y", d => scaleAndAxis.yScale(d.y) + lbl_padding_y)
+        }
+    }
+
+    createScatter(svg, box_width, box_height, points, dataPoints, scaleAndAxis) {
```

Comment:
this function is really really long. It needs either sub-sections inside of the body, or divide it to smaller helper init functions